### DISCUSSION TITLE
Is it possible to get the TaskInstance try_number in an Operator Extra Link?

### DISCUSSION THREAD

**Discussion by devstein** [apache/airflow#17659] [Q&A] [answered]
As per the title, we have a custom operator extra link that requires the task instance try number to correctly render the link. Is there a way to get this information? 

The current psuedo-code looks like this
```python
class TryNumberLink(BaseOperatorLink):
    name = 'TryNumberExample'

    def get_link(self, operator, dttm):
        ti = TaskInstance(task=operator, execution_date=dttm)
        # We need the task instance try number to be correct here 
        ti.render_templates()
        return ti.task.custom_template_var_with_try_number
```

**Comment by josh-fell** [comment]
Yes but you'll have to query the Airflow metastore backend to get the `TaskInstance` object and go from there.  Just keep in mind that querying the metastore backend will occur each time the Task Instance menu is revealed n the UI to generate the proper URL for the link.

For example, something like this should give you a link to the log of the latest task run based on the `try_number`. The psuedocode snippet provided would be creating a new `TaskInstance` of which the `try_number` would never change from its initial value. Querying the metastore should help.
```python
from airflow.utils.session import provide_session

class LastRunLogLink(BaseOperatorLink):
    name = 'Last Run Log'

    @provide_session
    def get_link(self, operator, dttm, session=None):
        tis = (
            session.query(TaskInstance)
            .filter(
                TaskInstance.dag_id == operator.dag_id,
                TaskInstance.task_id == operator.task_id,
                TaskInstance.execution_date == dttm,
            )
            .all()
        )
        ti = tis[0]
        try_number = max(1, ti.try_number - 1)
        return f"{ti.log_url}#{try_number}"
```

**Reply by devstein** [reply]
Thanks so much @josh-fell! I'll try this out. Do you have any advice for testing or debugging custom operator links?

**Reply by josh-fell** [reply]
You're welcome, let me know how it goes!

A great way to dive into testing these links -- or frankly anything in Airflow -- is to look at the unit tests out there.  Take a look at these unit tests: https://github.com/apache/airflow/blob/main/tests/www/views/test_views_extra_links.py

If you're talking about functional testing, for local development the open-source [Astronomer CLI](https://www.astronomer.io/docs/cloud/stable/resources/cli-reference) is a great way to get started (and full transparency I am an Airflow Ecosystem Engineer at Astronomer) or if you are thinking about making changes directly on a fork of Airflow, [Breeze](https://github.com/apache/airflow/blob/main/BREEZE.rst) is also a remarkably useful tool and is packaged in Airflow already.

Both tools will give you a nicely-packaged way of having complete control over the webserver locally so you can stop/restart at-will for testing.

**Reply by devstein** [reply]
For future viewers, here is the final solution. I relied on the `refresh_from_db` function rather than query the Metastore manually. 

```python
class CommuterLink(BaseOperatorLink):
    name = 'Commuter'

    @provide_session
    def get_link(self, operator, dttm, session=None):
        ti = TaskInstance(task=operator, execution_date=dttm)
        try:
            # Get task instances from Airflow metastore
            ti.refresh_from_db(session)
            # https://github.com/apache/airflow/blob/1.10.15/airflow/models/taskinstance.py#L215-L245
            # try_number will return the next try number (+1) unless the task is running,
            # then it returns the currently running task number
            # This updates the value, so it links to the last try attempt if the task is not running, 
            # else links to the currently running task
            ti.try_number = (
                ti._try_number - 1 if ti.state != State.RUNNING else ti._try_number
            )
        except Exception as e:
            # Print for debugging
            print(
                f"CommuterLink: error querying Airflow metastore for latest task number: {e}"
            )
        # Default to showing the first run's commuter link
        ti.render_templates()
        return ti.task.output_nb_commuter_url
```

Thanks for all the help @josh-fell 🙌

**Comment by devstein** [comment, marked as answer]
For future viewers, here is the final solution. I relied on the `refresh_from_db` function rather than query the Metastore manually. 

```python
class CommuterLink(BaseOperatorLink):
    name = 'Commuter'

    @provide_session
    def get_link(self, operator, dttm, session=None):
        ti = TaskInstance(task=operator, execution_date=dttm)
        try:
            # Get task instances from Airflow metastore
            ti.refresh_from_db(session)
            # https://github.com/apache/airflow/blob/1.10.15/airflow/models/taskinstance.py#L215-L245
            # try_number will return the next try number (+1) unless the task is running,
            # then it returns the currently running task number
            # This updates the value, so it links to the last try attempt if the task is not running, 
            # else links to the currently running task
            ti.try_number = (
                ti._try_number - 1 if ti.state != State.RUNNING else ti._try_number
            )
        except Exception as e:
            # Print for debugging
            print(
                f"CommuterLink: error querying Airflow metastore for latest task number: {e}"
            )
        # Default to showing the first run's commuter link
        ti.render_templates()
        return ti.task.output_nb_commuter_url
```

Thanks for all the help @josh-fell 🙌
